python/cucim/src/cucim/core/operations/color/jitter.py

```diff
@@ -0,0 +1,300 @@
+                C, H, W = input_arr.shape
+                N = 1
+
+            pitch = W * H
+            block = (128, 1, 1)
+            grid = (int((pitch - 1) / block[0] + 1), N, 1)
+            output_rgb = cupy.empty(input_arr.shape, dtype=cupy.uint8)
+            kernel_huejitter = CUDA_KERNELS.get_function("huejitter_kernel")
+            kernel_huejitter(grid, block, args=(input_arr,
+                                                output_rgb,
+                                                np.int32(pitch),
+                                                np.float32(hue)))
+
+            return output_rgb
+
+        # execution
+        f_brightness = check_input(brightness, 'brightness')
+        f_contrast = check_input(contrast, 'contrast')
+        f_saturation = check_input(saturation, 'saturation')
+        f_hue = check_input(hue, 'hue', center=0, bound=(-0.5, 0.5),
+                            clip_first_on_zero=False)
+
+        to_cupy = False
+
+        if isinstance(img, np.ndarray):
+            to_cupy = True
+            cupy_img = cupy.asarray(img, dtype=cupy.uint8, order="C")
+        elif not isinstance(img, cupy.ndarray):
+            raise TypeError("img must be a cupy.ndarray or numpy.ndarray")
+        else:
+            cupy_img = cupy.ascontiguousarray(img)
+
+        if cupy_img.dtype != cupy.uint8:
+            if cupy.can_cast(cupy_img.dtype, cupy.uint8, 'unsafe') is False:
+                raise ValueError(
+                    "Cannot cast type {cupy_img.dtype.name} to 'uint8'"
+                )
+            else:
+                cupy_img = cupy_img.astype(cupy.uint8)
+
+        if img.ndim not in (3, 4):
+            raise ValueError(
+                f"Unsupported img.ndim={img.ndim}. Expected `img` with "
+                "dimensions (C, H, W) or (N, C, H, W)."
+            )
+
+        fn_idx, brightness_factor, contrast_factor, saturation_factor, \
+            hue_factor = get_params(f_brightness, f_contrast,
+                                    f_saturation, f_hue)
+
+        for fn_id in fn_idx:
+            if fn_id == 0 and brightness_factor is not None:
+                cupy_img = adjust_brightness(cupy_img, brightness_factor)
+            elif fn_id == 1 and contrast_factor is not None:
+                cupy_img = adjust_contrast(cupy_img, contrast_factor)
+            elif fn_id == 2 and saturation_factor is not None:
+                cupy_img = adjust_saturation(cupy_img, saturation_factor)
+            elif fn_id == 3 and hue_factor is not None:
+                cupy_img = adjust_hue(cupy_img, hue_factor)
+
+        if img.dtype != np.uint8:
+            cupy_img = cupy_img.astype(cupy.float32)
+
+        result = cupy_img
+        if to_cupy is True:
+            result = cupy.asnumpy(cupy_img)
+
+        return result
+    except Exception as e:
+        _logger.error("[cucim] " + str(e), exc_info=True)
+        _logger.info("Error executing color jitter on GPU")
+        raise
```

Comment:
Do we want to handle logging within cuCIM? Typically this is the sort of thing that would be configured in a user application (unless for some reason the exception would never be seen)

Reply:
@shekhardw, any thoughts on this?

Reply:
I propose to fix this in a separate MR.